django/contrib/auth/checks.py

```diff
@@ -52,7 +53,11 @@ def check_user_model(app_configs=None, **kwargs):
         )
 
     # Check that the username field is unique
-    if not cls._meta.get_field(cls.USERNAME_FIELD).unique:
+    if not cls._meta.get_field(cls.USERNAME_FIELD).unique and not any(
+        isinstance(constraint, models.UniqueConstraint) and
+        constraint.fields == (cls.USERNAME_FIELD,)
+        for constraint in cls._meta.constraints
+    ):
```

Comment:
I think you can to exclude conditional unique constraints here

```suggestion
    if not cls._meta.get_field(cls.USERNAME_FIELD).unique and not any(
        isinstance(constraint, models.UniqueConstraint) and
        constraint.fields == (cls.USERNAME_FIELD,)
        for constraint in cls._meta.total_unique_constraints
    ):
```

Kind of wish there was a way to avoid checking both the field and `cls._meta.total_unique_constraints`.

Reply:
Yeah that probably makes sense. I guess one could implement a conditional unique constraint to ignore soft deleted users but maybe that warrant an exclusion of this check by adding it to `SILENCED_SYSTEM_CHECKS`.